### ISSUE TITLE
request: new function rectangular matrix determinant

### ISSUE THREAD

**Issue by latot** [sympy/sympy#10318]
Hi all, first happy day.

Well i wanna request this new function to sympy, i don't use python but i works with octave and symbolic package, it uses sympy (layer octave/sympy).

well i know strict talking non-square matrix don't have determinant but actually it have a solution with some conditions (columns>=rows) and with the jacobian interpretation, one of the reazon to why i think it should be implemented is, ppl don't use if don't see it or don't know it exist, so although can be very easy formula i think can help here.

formula:

```
det(m)=det(m*m')^(1/2)
Conditions: Columns=>Rows, Always Positive Solution (or think it as absolute value)
m' transpose of m
```

Well with jacobian interpretation i mean the result its the representation of the space betwen vectors, ex, if we have 2 vectors in 3 dimentions, both can be rotated to two dimentions and recalculated the determinant to get the space, this is a simplification of that, its very ussefull and a super advantaje its it works with multi-dimentional equations, actually we use cross option to get this, but it have this limitations (3 dims), now a practical example (in octave and without complex nmbers)(det funcion overloaded with up formula, call it as you like):

This is to get the area of an sphere, we make the matrix with the differencials.
```
syms r a b
x=r*sin(a)*cos(b)
y=r*sin(a)*sin(b)
z=r*cos(a)
u=[diff(x, a) diff(y, a) diff(z, a);diff(x, b) diff(y, b) diff(z, b)]
u = (sym 2×3 matrix)

  ⎡r⋅cos(a)⋅cos(b)   r⋅sin(b)⋅cos(a)  -r⋅sin(a)⎤
  ⎢                                            ⎥
  ⎣-r⋅sin(a)⋅sin(b)  r⋅sin(a)⋅cos(b)      0    ⎦

#last line is [0 0 0] diff on constant r
octave:58> simplify(det(u))
warning: Using rat() heuristics for double-precision input (is this what you wanted?)
warning: called from
    sym at line 221 column 7
    detm at line 5 column 2
ans = (sym)

     ____________
    ╱  4    2    
  ╲╱  r ⋅sin (a) 
```

As you can see you get the same value of the the normal jacobian calculation and this method.

Docs:
https://en.wikipedia.org/wiki/Determinant
check Volume and Jacobian determinant

https://www.math.uchicago.edu/~may/VIGRE/VIGRE2007/REUPapers/FINALAPP/Peng.pdf

Cya.

**Comment by shaoron** [comment]
Jacobian function exists. You can type
u = Matrix([x,y,z]).jacobian([a,b])
simplify(sqrt(det(u.T*u)))

Output is the same.

I can write something like "jacobian_det", if it is needed.

**Comment by latot** [comment]
Hi, willl be greate, but i have a little question, what shoud be used?
```
simplify(sqrt(det(u.T*u)))
```
or
```
simplify(abs(sqrt(det(u.T*u))))
```
this depends how works sympy, if the las sqrt don't is removed and is used as always positive solution we can skip absolute value.
Note this should works with more than 3 dimentions.

**Comment by shaoron** [comment]
Should it work as
Matrix(list of vectors).jacobian_det(list of dimentions)
OR
jacobian_det(Matrix)?

First one is clearer, I guess.

**Comment by latot** [comment]
Hi, seems fine to me use the first, some little things, should be:
```
if mat.cols == mat.rows:
 return mat.det()
elif mat.cols >= mat.rows:
 return sqrt((mat*mat.T).det()) 
else:
 return sqrt((mat.T*mat).det())
```

and only to be sure about sqrt, it never sould be simplified even if we have cuadratic or somethings else.

Thx.

**Comment by shaoron** [comment]
Ok, thanks for good explanation.

sqrt works as you wish, I guess

**Comment by latot** [comment]
be careful:
```
+        mat = self.jacobian(X)
 +        if mat.cols == mat.rows:
 +            return mat.det()
 +        elif mat.cols > mat.rows:
 +            return sqrt((mat*mat.T).det())
 +        else:
 +            return sqrt((mat*mat.T).det())
```

last line should be:
```
 +            return sqrt((mat.T*mat).det())
```

**Comment by latot** [comment]
Hi, sadly we need do other change, i test in sympy this:
```
sqrt(x)*sqrt(x)
```
and finally the sqrt is removed D:

this:
```
        if mat.cols == mat.rows:
             return mat.det()
         elif mat.cols > mat.rows:
             return sqrt((mat*mat.T).det())
         else:
             return sqrt((mat.T*mat).det())
```
now should be:
```
        if mat.cols == mat.rows:
             return abs(mat.det())
         elif mat.cols > mat.rows:
             return abs(sqrt((mat*mat.T).det()))
         else:
             return abs(sqrt((mat.T*mat).det()))
```

**Comment by latot** [comment]
https://github.com/shaoron/sympy/pull/1

**Comment by shaoron** [comment]
Sorry, exams.

Docs:
Sqrt always return positive solution, for example simplify(sqrt(x**2)) == sqrt(x**2)
But simplify(sqrt(x)**2) ==  x

We do not need abs

**Comment by latot** [comment]
yes we need it, in your last example should be the positive solution, the jacobian determinant for non-square matrix and for square matrix always use abs, in other words, volume is always positive even if is ^2. 

other doc:
http://tutorial.math.lamar.edu/Classes/CalcIII/ChangeOfVariables.aspx

here show how is used abs to calculate jacobian determinant.
